pkg/logging/logging.go

```diff
@@ -0,0 +1,182 @@
+	defaultLogLevel, ok := logFilterLevels["info"]
+	if !ok {
+		defaultLogLevel = 10
+	}
+	return &FilteredLogger{logger: logger,
+		component:             defaultComponent,
+		filterLevel:           defaultLogLevel,
+		currentLogLevel:       defaultLogLevel,
+		verbosityLevel:        defaultVerbosity,
+		currentVerbosityLevel: defaultVerbosity,
+	}
+}
+
+type NullLogger struct{}
+
+func (n NullLogger) Log(params ...interface{}) error { return nil }
+
+var loggers = make(map[string]*FilteredLogger)
+var initialized = false
+var defaultComponent = ""
+var defaultVerbosity = 0
+
+func Logger(component string) *FilteredLogger {
+	if _, ok := loggers[component]; !ok {
+		logger := log.NewLogfmtLogger(os.Stderr)
+		log := MakeLogger(logger)
+		log.component = component
+		loggers[component] = log
+	}
+	return loggers[component]
+}
+
+func (l *FilteredLogger) SetLogger(logger log.Logger) *FilteredLogger {
+	l.logger = logger
+	return l
+}
+
+var logFilterLevels = map[string]int{
+	"debug":    0,
+	"info":     10,
+	"warning":  20,
+	"error":    30,
+	"critical": 40}
+
+type LogError struct {
+	message string
+}
+
+func (e LogError) Error() string {
+	return e.message
+}
+
+func (l FilteredLogger) Log(params ...interface{}) error {
+	// messages should be logged if any of these conditions are met:
+	// The log filtering level is debug
+	// The log filtering level is info and verbosity checks match
+	// The log message priority is warning or higher
+	force := (l.filterLevel == logFilterLevels["debug"]) || (l.currentLogLevel >= logFilterLevels["warning"])
+
+	if force || (l.filterLevel == logFilterLevels["info"] &&
+		(l.currentLogLevel == l.filterLevel) &&
+		(l.currentVerbosityLevel <= l.verbosityLevel)) {
+		logParams := make([]interface{}, 0)
+		logParams = append(logParams, "component", l.component)
+
+		if l.currentVerbosityLevel > 1 {
+			_, fileName, lineNumber, _ := runtime.Caller(1)
+			logParams = append(logParams, "filename", fileName)
+			logParams = append(logParams, "line", lineNumber)
+		}
+		logParams = append(logParams, params...)
+		return l.logger.Log(logParams...)
+
+	}
+	return nil
+}
+
+func (l FilteredLogger) Object(obj ...interface{}) error {
+	// name, ID, type
+	t := reflect.TypeOf(obj).Elem()
+	logParams := make([]interface{}, 0)
+	logParams = append(logParams, "component", l.component)
+
+	if l.currentVerbosityLevel > 0 {
+		_, fileName, lineNumber, _ := runtime.Caller(1)
+		logParams = append(logParams, "filename", fileName)
+		logParams = append(logParams, "line", lineNumber)
+	}
+
+	logParams = append(logParams, "type", t.String())
+	logParams = append(logParams, "name", t.Name())
+
+	if (t.Kind() == reflect.Array) || (t.Kind() == reflect.Slice) {
+		logParams = append(logParams, "length", t.Len())
+	}
+	if t.Kind() == reflect.Struct {
+		logParams = append(logParams, "fields", t.NumField())
+	}
+
+	return l.logger.Log(logParams)
+}
+
+func (l *FilteredLogger) SetLogLevel(filter string) error {
```

Comment:
You can create a string subtype to make it typesave. As a consequence, only valid filter values can be passed in:

```golang
type LogLevel string
const (
DEBUG LogLevel = "debug"
...
)
func (l *FilteredLogger) SetLogLevel(filter LogLevel) {
l.filterLevel = filter
}
```